api/controllers/grades.go

```diff
@@ -6,283 +6,202 @@ import (
-		var courseMatch bson.D
-		var courseFind bson.D
-		var professorMatch bson.D
-		var professorFind bson.D
+	unwindGradeDistributionStage := bson.D{{Key: "$unwind", Value: "$grade_distribution"}}
 
-		var err error
+	groupGradesStage := bson.D{
+		{Key: "$group", Value: bson.D{
+			{Key: "_id", Value: bson.D{
+				{Key: "academic_session", Value: "$_id"},
+				{Key: "ix", Value: "$ix"},
+			}},
+			{Key: "grades", Value: bson.D{{Key: "$push", Value: "$grade_distribution"}}},
+		}},
+	}
 
-		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
-		defer cancel()
+	sortGradesStage := bson.D{
+		{Key: "$sort", Value: bson.D{
+			{Key: "_id.ix", Value: 1},
+			{Key: "_id", Value: 1},
+		}},
+	}
 
-		// @TODO: Recommend forcing using first_name and last_name to ensure single professors per query.
-		// All professors sharing the name will be aggregated together in the current implementation
-		prefix := c.Query("prefix")
-		number := c.Query("number")
-		section_number := c.Query("section_number")
-		first_name := c.Query("first_name")
-		last_name := c.Query("last_name")
+	sumGradesStage := bson.D{{Key: "$addFields", Value: bson.D{{Key: "grades", Value: bson.D{{Key: "$sum", Value: "$grades"}}}}}}
 
-		professor := (first_name != "" || last_name != "")
+	groupGradeDistributionStage := bson.D{
+		{Key: "$group", Value: bson.D{
+			{Key: "_id", Value: "$_id.academic_session"},
+			{Key: "grade_distribution", Value: bson.D{{Key: "$push", Value: "$grades"}}},
+		}},
+	}
 
-		lookupSectionsStage := bson.D{
-			{Key: "$lookup", Value: bson.D{
-				{Key: "from", Value: "sections"},
-				{Key: "localField", Value: "sections"},
-				{Key: "foreignField", Value: "_id"},
-				{Key: "as", Value: "sections"},
-			}},
+	switch {
+	case prefix != "" && number == "" && sectionNumber == "" && !professor:
+		// Filter on Course
+		collection = courseCollection
+		pipeline = append(pipeline, bson.D{{Key: "$match", Value: bson.M{"subject_prefix": prefix}}})
+
+	case prefix != "" && number != "" && sectionNumber == "" && !professor:
+		// Filter on Course
+		collection = courseCollection
+		pipeline = append(pipeline, bson.D{{Key: "$match", Value: bson.M{"subject_prefix": prefix, "course_number": number}}})
+
+	case prefix != "" && number != "" && sectionNumber != "" && !professor:
+		// Filter on Course then Section
+		collection = courseCollection
+		pipeline = append(pipeline, bson.D{{Key: "$match", Value: bson.M{"subject_prefix": prefix, "course_number": number}}})
+		pipeline = append(pipeline, bson.D{{Key: "$match", Value: bson.M{"sections.section_number": sectionNumber}}})
+
+	case prefix == "" && number == "" && sectionNumber == "" && professor:
+		// Filter on Professor
+		collection = professorCollection
+		match := bson.M{}
+		if firstName != "" {
+			match["first_name"] = firstName
+		}
+		if lastName != "" {
+			match["last_name"] = lastName
 		}
+		pipeline = append(pipeline, bson.D{{Key: "$match", Value: match}})
```

Comment:
This code isn't doing the same thing that the old code was doing, so will work improperly; most of the pipeline stages end up missing!

This actually causes a build error due to most of the stage variables being unused, and is something that should be caught during local testing! Make sure you're building with `build.bat` on Windows or with the `makefile` on Mac/Linux to catch issues like this.